File: erlang/trie/src/trie.erl
-module(trie).

%% API exports
-export([new/0, insert/3, lookup/2, traversal/1, traversal_limit/2, expand/3,
        scan_content/2]).

-define(TRIE, ?MODULE).

-record(?TRIE, {children,
                data:: term()}).

-type tree() :: #?TRIE{}.

%%====================================================================
%% API functions
%%====================================================================

-spec new() -> tree().
new() ->
    #{children => #{}, data => nil}.

-spec insert(tree(), string(), term()) -> tree().
insert(Trie, [], Data) ->
    Trie#{data => Data};
insert(Trie, [K1 | Kpai], Data) ->
    Child = maps:get(children, Trie),
    case maps:find(K1, Child) of
        {ok, Value} ->
            Trie#{children => maps:put(K1, insert(Value, Kpai, Data), Child)};
        error ->
            Trie#{children => maps:put(K1, insert(new(), Kpai, Data), Child)}
    end.

-spec lookup(tree(), string()) -> term().
lookup(Trie, []) ->
    maps:get(data, Trie);
lookup(Trie, [K1 | Kpai]) ->
    Child = maps:get(children, Trie),
    case maps:find(K1, Child) of
        {ok, Value} ->
            lookup(Value, Kpai);
        error ->
            undefined
    end.

-spec traversal(tree()) -> list().
traversal(Trie) ->
    traversal(Trie, "", "").


-spec traversal_limit(tree(), number()) -> list().
traversal_limit(Trie, MaxCnt) ->
    traversal_limit(Trie, "", "", MaxCnt, 0).

-spec expand(tree(), string(), number()) -> list().
expand(Trie, Prefix, N) ->
    expand(Trie, Prefix, N, Prefix).


- spec scan_content(string(), tree()) -> list().
scan_content(Content, Trie) ->
    case Content of
        [] -> [];
        [H | T] -> lists:reverse(scan_content([H], T, 0, Trie, [], 0))
    end.

%%====================================================================
%% Internal functions
%%====================================================================

traversal(Trie, Path, Kvs) ->
    Data = maps:get(data, Trie),
    case map_size(maps:get(children, Trie)) of
        0 ->
            case Data of
                nil -> Kvs;
                _ -> Kvs ++ [{Path, Data}]
            end;
        _ ->
            case Data of
                nil -> Kvs2 = Kvs;
                _ ->
                    Kvs2 = [{Path, Data} | Kvs]
            end,
            Children = maps:get(children, Trie),
            lists:foldl(fun(Child, Ret) ->
                                Ret ++ traversal(maps:get(Child, Children), Path ++ [Child], "")
                        end, Kvs2, maps:keys(Children))
    end.


traversal_limit(_, _, Kvs, MaxCnt, Cnt) when Cnt >= MaxCnt ->
    {Kvs, Cnt};
traversal_limit(Trie, Path, Kvs, MaxCnt, Cnt) ->
    Data = maps:get(data, Trie),
    case map_size(maps:get(children, Trie)) of
        0 ->
            case Data of
                nil -> {Kvs, Cnt};
                _ ->
                    {Kvs ++ [{Path, Data}], Cnt + 1}
            end;
        _ ->
            case Data of
                nil ->
                    Cnt2 = Cnt,
                    Kvs2 = Kvs;
                _ ->
                    Cnt2 = Cnt + 1,
                    Kvs2 = [{Path, Data} | Kvs]
            end,
            Children = maps:get(children, Trie),
            lists:foldl(fun(Child, Ret) ->
                                Next = traversal_limit(maps:get(Child, Children), Path ++ [Child], "", MaxCnt, erlang:element(2, Ret)),
                                {erlang:element(1, Ret) ++ erlang:element(1, Next), erlang:element(2, Next)}
                        end, {Kvs2, Cnt2}, maps:keys(Children))
    end.


expand(Trie, [], N, Prefix) ->
    add_prefix(traversal_limit(Trie, N), Prefix);
expand(Trie, [K1 | Kpai], N, Prefix) ->
    Child = maps:get(children, Trie),
    case maps:find(K1, Child) of
        {ok, Value} ->
            expand(Value, Kpai, N, Prefix);
        error ->
            undefined
    end.


add_prefix(Ret, Prefix) ->
    case Ret of
        undefined ->
            undefined;
        {Kvs, Cnt} ->
            Kvs2 = lists:map(fun({K, V}) ->
                                     {Prefix ++ K, V}
                             end, Kvs),
            {Kvs2, Cnt};
        _ ->
            undefined
    end.


scan_content(K, [], _, Trie, Ret, Offset) ->
    case lookup(Trie, K) of
        undefined ->
            Ret;
        nil ->
            Ret;
        Data ->
            [{{K, Data}, Offset} | Ret]
    end;
scan_content(K, Remain, OffsetRemain, Trie, Ret, Offset)
            when length(Remain) > OffsetRemain ->
    case lookup(Trie, K) of
        undefined ->
            [RemainH | RemainT] = Remain,
            scan_content([RemainH], RemainT, 0, Trie, Ret, Offset + 1);
        nil ->
            scan_content(K ++ [lists:nth(OffsetRemain + 1, Remain)], Remain,
                         OffsetRemain + 1, Trie, Ret, Offset);
        Data ->
            Ret2 = [{{K, Data}, Offset} | Ret],
            scan_content(K ++ [lists:nth(OffsetRemain + 1, Remain)], Remain,
                         OffsetRemain + 1, Trie, Ret2, Offset)
    end;
scan_content(K, Remain, _OffsetRemain, Trie, Ret, Offset) ->
    [RemainH | RemainT] = Remain,
    case lookup(Trie, K) of
        undefined ->
            scan_content([RemainH], RemainT, 0, Trie, Ret, Offset + 1);
        nil ->
            scan_content([RemainH], RemainT, 0, Trie, Ret, Offset + 1);
        Data ->
            Ret2 = [{{K, Data}, Offset} | Ret],
            scan_content([RemainH], RemainT, 0, Trie, Ret2, Offset + 1)
    end.

File: elixir/trie/lib/trie.ex
defmodule Trie do
  @moduledoc """
  Documentation for `Trie`.
  """

  @doc """
  Create the `Trie` data structure.

  ## Examples

      iex> Trie.new()
      %Trie{children: %{}, data: nil}

  """
  defstruct children: %{}, data: nil

  def new do
    %Trie{}
  end

  def insert(trie, [], data) do
    Map.put(trie, :data, data)
  end
  def insert(trie, [k1 | kpai], data) do
    child = trie[:children]
    case Map.has_key?(child, k1) do
      true -> Map.put(trie, k1, insert(child[k1], kpai, data))
      false -> Map.put(trie, k1, insert(new(), kpai, data))
    end
  end

  def lookup(trie, []) do
    trie[:data]
  end
  def lookup(trie, [k1 | kpai]) do
    child = trie[:children]
    case Map.has_key?(child, k1) do
      true -> lookup(child[k1], kpai)
      false -> :undefined
    end
  end

  def traversal(trie) do
    traversal(trie, '', '')
  end
  defp traversal(trie, path, kvs) do
    data = trie[:data]
    case map_size(trie[:children]) do
      0 ->
        case data do
          nil -> kvs
          _ -> kvs ++ [{path, data}]
        end
      _ ->
        children = trie[:children]
        case data do
          nil ->
            List.foldl(Map.keys(children), kvs, fn child, ret ->
              ret ++ traversal(children[child], path ++ [child], '')
            end)
          _ ->
            List.foldl(Map.keys(children), [{path, data} | kvs], fn child, ret ->
              ret ++ traversal(children[child], path ++ [child], '')
            end)
        end
    end
  end

  def traversal_limit(trie, max_cnt) do
    traversal_limit(trie, '', '', max_cnt, 0)
  end
  defp traversal_limit(_, _, kvs, max_cnt, cnt) when cnt >= max_cnt do
    {kvs, cnt}
  end
  defp traversal_limit(trie, path, kvs, max_cnt, cnt) do
    data = trie[:data]
    case map_size(trie[:children]) do
      0 ->
        case data do
          nil -> {kvs, cnt}
          _ -> {kvs ++ [{path, data}], cnt + 1}
        end
      _ ->
        children = trie[:children]
        case data do
          nil ->
            cnt2 = cnt; kvs2 = kvs
            List.foldl(Map.keys(children), {kvs2, cnt2}, fn child, ret ->
              next = traversal_limit(children[child], path ++ [child], '', max_cnt, :erlang.element(2, ret))
              {:erlang.element(1, ret) ++ :erlang.element(1, next), :erlang.element(2, next)}
            end)
          _ ->
            cnt2 = cnt + 1; kvs2 = [{path, data} | kvs]
            List.foldl(Map.keys(children), {kvs2, cnt2}, fn child, ret ->
              next = traversal_limit(children[child], path ++ [child], '', max_cnt, :erlang.element(2, ret))
              {:erlang.element(1, ret) ++ :erlang.element(1, next), :erlang.element(2, next)}
            end)
        end
    end
  end

  def expand(trie, prefix, n) do
    expand(trie, prefix, n, prefix)
  end

  defp add_prefix(ret, prefix) do
    case ret do
      :undefined ->
        :undefined
      {kvs, cnt} ->
        kvs2 = Enum.map(kvs, fn {k, v} -> {prefix ++ k, v} end)
        {kvs2, cnt}
      _ ->
        :undefined
    end
  end

  defp expand(trie, '', n, prefix) do
    add_prefix(traversal_limit(trie, n), prefix)
  end
  defp expand(trie, [k1 | kpai], n, prefix) do
    child = trie[:children]
    case Map.has_key?(child, k1) do
      true -> expand(child[k1], kpai, n, prefix)
      false -> :undefined
    end
  end

  def scan_content(content, trie) do
    case content do
      [] -> ''
      [H | T] -> Enum.reverse(scan_content([H], T, 0, trie, [], 0))
    end
  end
  defp scan_content(k, '', _, trie, ret, offset) do
    case lookup(trie, k) do
      :undefined -> ret
      nil -> ret
      data -> [{{k, data}, offset} | ret]
    end
  end
  defp scan_content(k, remain, offset_remain, trie, ret, offset) when length(remain) > offset_remain do
    case lookup(trie, k) do
      :undefined ->
        [remain_h | remain_t] = remain
        scan_content([remain_h], remain_t, 0, trie, ret, offset + 1)
      nil ->
        scan_content(k ++ [Enum.at(remain, offset_remain)], remain, offset_remain + 1, trie, ret, offset)
      data ->
        ret2 = [{{k, data}, offset} | ret]
        scan_content(k ++ [Enum.at(remain, offset_remain)], remain, offset_remain + 1, trie, ret2, offset)
    end
  end
  defp scan_content(k, remain, _offset_remain, trie, ret, offset) do
    [remain_h | remain_t] = remain
    case lookup(trie, k) do
      :undefined ->
        scan_content([remain_h], remain_t, 0, trie, ret, offset + 1)
      nil ->
        scan_content([remain_h], remain_t, 0, trie, ret, offset + 1)
      data ->
        ret2 = [{{k, data}, offset} | ret]
        scan_content([remain_h], remain_t, 0, trie, ret2, offset + 1)
    end
  end
end
